runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,740 @@
+
+        /// Salt and referendum option provided don't correspond to the commitment
+        InvalidReveal,
+
+        /// Vote for not existing option was revealed
+        InvalidVote,
+
+        /// Trying to reveal vote that was not casted
+        NoVoteToReveal,
+    }
+}
+
+impl<T: Trait<I>, I: Instance> PartialEq for Error<T, I> {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_u8() == other.as_u8()
+    }
+}
+
+impl<T: Trait<I>, I: Instance> From<BadOrigin> for Error<T, I> {
+    fn from(_error: BadOrigin) -> Self {
+        Error::<T, I>::BadOrigin
+    }
+}
+
+/////////////////// Module definition and implementation ///////////////////////
+
+decl_module! {
+    pub struct Module<T: Trait<I>, I: Instance> for enum Call where origin: T::Origin {
+        /// Predefined errors
+        type Error = Error<T, I>;
+
+        /// Setup events
+        fn deposit_event() = default;
+
+        /////////////////// Lifetime ///////////////////////////////////////////
+
+        // start voting period
+        #[weight = 10_000_000]
+        pub fn start_referendum(origin, options: Vec<T::ReferendumOption>, winning_target_count: u64) -> Result<(), Error<T, I>> {
+            // ensure action can be started
+            EnsureChecks::<T, I>::can_start_referendum(origin, &options)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            // update state
+            Mutations::<T, I>::start_voting_period(&options, &winning_target_count);
+
+            // emit event
+            Self::deposit_event(RawEvent::ReferendumStarted(options, winning_target_count));
+
+            Ok(())
+        }
+
+        // finish voting period
+        #[weight = 10_000_000]
+        pub fn finish_voting_start_revealing(origin) -> Result<(), Error<T, I>> {
+            // ensure action can be started
+            EnsureChecks::<T, I>::can_finish_voting(origin)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            // start revealing phase
+            Mutations::<T, I>::start_revealing_period();
+
+            // emit event
+            Self::deposit_event(RawEvent::RevealingStageStarted());
+
+            Ok(())
+        }
+
+        #[weight = 10_000_000]
+        pub fn finish_revealing_period(origin) -> Result<(), Error<T, I>> {
+            // ensure action can be started
+            EnsureChecks::<T, I>::can_finish_revealing(origin)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            // start revealing phase
+            let referendum_result = Mutations::<T, I>::conclude_referendum();
+
+            // emit event
+            Self::deposit_event(RawEvent::ReferendumFinished(referendum_result));
+
+            Ok(())
+        }
+
+        /////////////////// User actions ///////////////////////////////////////
+
+        #[weight = 10_000_000]
+        pub fn vote(origin, commitment: T::Hash, stake: T::CurrencyBalance) -> Result<(), Error<T, I>> {
```

Comment:
The spec clearly describes how voting involves specifying in detail what accounts should be drawn upon for how much stake. It's also not clear where the locks are being applied for the actual voting, this is required, otherwise what would prevent different people voting with the same stake again and again? `can_vote` cannot do this without mutating locks, which is not what it should be doing. Lastly, be sure in the fixed version that the council module excludes council candidates from being able to vote in the implementation of this.

You are also not doing any checks as to whether there is sufficient free stake to do the voting. The call to `can_vote` cannot do any enforcement here, because it would have to update locks to reflect the fact that an actual vote had been made

Reply:
Reworked.